add new command kubectl set volume

WanLinghao opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:
#21648 
Moved from OpenShift to Kubenetes.
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```
Add 'kubectl set volume' command for setting volume in containers for resources embedding pod templates
```

CaoShuFeng commented:
/ok-to-test

mkumatag commented:
`api.Codecs` moved to `legacyscheme.Codecs` part of https://github.com/kubernetes/kubernetes/pull/53984, so you may need to fix the import path to build the code.

cblecker commented:
/unassign
Please reassign back if you need `hack/` approval once this has been reviewed by a sig-cli reviewer :)

WanLinghao committed:
this patch add "kubectl set volume" for easier updating container volumes
	modified:   docs/.generated_docs
	new file:   docs/man/man1/kubectl-set-volumes.1
	new file:   docs/user-guide/kubectl/kubectl_set_volumes.md
	modified:   hack/make-rules/test-cmd-util.sh
	modified:   pkg/kubectl/cmd/set/set.go
	new file:   pkg/kubectl/cmd/set/set_volume_test.go
	new file:   pkg/kubectl/cmd/set/set_volume.go
	modified:   pkg/kubectl/cmd/set/BUILD

kargakis commented:
@kubernetes/sig-cli-pr-reviews

WanLinghao commented:
/retest pull-kubernetes-unit

WanLinghao commented:
/test pull-kubernetes-unit

dixudx commented on pkg/kubectl/cmd/set/set_volume.go:
```diff
@@ -0,0 +1,958 @@
+func (o *VolumeOptions) addVolumeToSpec(spec *api.PodSpec, info *resource.Info, singleResource bool) error {
+	opts := o.AddOpts
+	if len(o.Name) == 0 {
+		var err error
```
Better change to

```
if o.Name, err := o.getVolumeName(spec, singleResource); err != nil {
	return err
}
```

WanLinghao replied:
I think cause o.Name is an existed variable, so your way is not suitable

WanLinghao replied:
```
var err error
if o.Name, err = o.getVolumeName(spec, singleResource); err != nil {
	return err
}

```
did you mean this?

dixudx replied:
SGTM

dixudx commented on pkg/kubectl/cmd/set/set_volume.go:
```diff
@@ -0,0 +1,958 @@
+func (o *VolumeOptions) addVolumeToSpec(spec *api.PodSpec, info *resource.Info, singleResource bool) error {
+	opts := o.AddOpts
+	if len(o.Name) == 0 {
+		var err error
```
Better change to

```
if o.Name, err := o.getVolumeName(spec, singleResource); err != nil {
	return err
}
```

dixudx replied:
Better change to 

```
if o.Name, err := o.getVolumeName(spec, singleResource); err != nil {
	return err
}
```

WanLinghao commented:
/assign cblecker

dixudx commented on pkg/kubectl/cmd/set/set_volume.go:
```diff
@@ -0,0 +1,958 @@
+			return errors.New("either specify --type or --source but not both for --add operation")
+		}
+
+		if len(a.Type) > 0 {
```
I wonder shall we split these checking to separate methods to make it more concise and maintainable.

WanLinghao replied:
@dixudx OK

dixudx commented on pkg/kubectl/cmd/set/set_volume.go:
```diff
@@ -0,0 +1,958 @@
+func (a *AddVolumeOptions) Validate(isAddOp bool) error {
+	if isAddOp {
+		if len(a.Type) == 0 && (len(a.ClaimName) > 0 || len(a.ClaimSize) > 0) {
+			a.Type = "persistentvolumeclaim"
```
Actually I don't like such a hard-coded way.

WanLinghao replied:
@dixudx please make it more clear.Do you mean this function contains too much things?

WanLinghao commented on pkg/kubectl/cmd/set/set_volume.go:
```diff
@@ -0,0 +1,958 @@
+
+const (
+	volumePrefix    = "volume-"
+	storageAnnClass = "volume.beta.kubernetes.io/storage-class"
```
OK

shiywang commented on pkg/kubectl/cmd/set/set_volume.go:
```diff
@@ -0,0 +1,960 @@
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apiserver/pkg/storage/names"
+	"k8s.io/kubernetes/pkg/api"
```
no internal type

WanLinghao replied:
@shiywang  I am not clear with version stuff,  would please give me some hints, thanks a lot

shiywang commented on pkg/kubectl/cmd/set/set_volume.go:
```diff
@@ -0,0 +1,960 @@
+
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	kresource "k8s.io/apimachinery/pkg/api/resource"
```
nit: i would prefer just "k8s.io/apimachinery/pkg/api/resource" since there's no upstream dependency anymore.

WanLinghao replied:
since this file import another resource package   "k8s.io/kubernetes/pkg/kubectl/resource"

shiywang commented on pkg/kubectl/cmd/set/set_volume.go:
```diff
@@ -0,0 +1,960 @@
+}
+
+func (o *VolumeOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, out, errOut io.Writer) error {
+	kc, err := f.ClientSet()
```
in kubernetes, we can just call it client

shiywang commented on pkg/kubectl/cmd/set/set_volume.go:
```diff
@@ -0,0 +1,960 @@
+		if len(o.AddOpts.ClaimName) == 0 {
+			o.AddOpts.ClaimName = names.SimpleNameGenerator.GenerateName("pvc-")
+		}
+		q, err := kresource.ParseQuantity(o.AddOpts.ClaimSize)
```
ditto

zjj2wry replied:
+1

shiywang commented:
just have a glance, will take a deeper look tomorrow morning

zjj2wry commented on pkg/kubectl/cmd/set/set_volume_test.go:
```diff
@@ -0,0 +1,282 @@
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/kubernetes/pkg/api"
```
this should be removed, use versioned api

WanLinghao replied:
use "k8s.io/client-go/kubernetes/typed/core/v1" replacing  "k8s.io/kubernetes/pkg/api"?

WanLinghao replied:
@zjj2wry since I am still not clear with this version stuff

zjj2wry replied:
api rep: `k8s.io/api...`
api maybe in: `k8s.io/api/core/v1`

WanLinghao replied:
@zjj2wry  Is it a new move still on pushing? I notice that file in same package like set_env.go, set_image.go still use "k8s.io/kubernetes/pkg/api". 
since connections between them exist, if use core/v1 in set_volume.go. then it is necessary that  we should use the same in this whole package

zjj2wry replied:
yes, will fix all set command :), and it is in process

WanLinghao replied:
@zjj2wry so would mind if I just ignore this right now and fix it after other set command?
by the way , are there any PR that focus on this? please tell me so I can keep in track

liggitt replied:
it's being actively worked on in https://github.com/kubernetes/kubernetes/pull/54554

@WanLinghao I would prefer not adding this until that PR is in

WanLinghao replied:
@liggitt  OK, I will wait for this and rebase then

zjj2wry commented on pkg/kubectl/cmd/set/set_volume_test.go:
```diff
@@ -0,0 +1,282 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/fake"
```
ditto

zjj2wry commented on pkg/kubectl/cmd/set/set_volume.go:
```diff
@@ -0,0 +1,960 @@
+	"k8s.io/apiserver/pkg/storage/names"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+	kcoreclient "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/internalversion"
```
should be versioned

zjj2wry commented on pkg/kubectl/cmd/set/set_volume.go:
```diff
@@ -0,0 +1,960 @@
+		* persistentvolumeclaim or pvc (persistent volume claim): Link the volume
+		  directory in the container to a persistent volume claim you have allocated by
+			name - a persistent volume claim is a request to allocate storage. Note that
+			if your claim hasn't been bound, your pods will not start.
```
this format is on purpose?

WanLinghao replied:
this format was imported from original file, I will keep it  same with k8s style

zjj2wry commented on pkg/kubectl/cmd/set/set_volume.go:
```diff
@@ -0,0 +1,960 @@
+		* secret (mounted secret): Secret volumes mount a named secret to the provided
+		  directory.
+
+		For descriptions on other volume types, see https://docs.openshift.com`)
```
should be use kubernetes.IO not openshift

WanLinghao replied:
what about just delete this line

zjj2wry commented on pkg/kubectl/cmd/set/set_volume.go:
```diff
@@ -0,0 +1,960 @@
+			SelectorParam(o.Selector).
+			ResourceTypeOrNameArgs(o.All, args...).
+			Latest()
+	} else {
```
```
		// if a --local flag was provided, and a resource was specified in the form
		// <resource>/<name>, fail immediately as --local cannot query the api server
		// for the specified resource.
		if len(o.Resources) > 0 {
			return resource.LocalResourceError
		}
```
should be check when --local and resources both special

zjj2wry commented on pkg/kubectl/cmd/set/set_volume.go:
```diff
@@ -0,0 +1,960 @@
+}
+
+func (o *VolumeOptions) RunVolume(args []string, f cmdutil.Factory) error {
+	b := f.NewBuilder().
```
use build will better instead b

zjj2wry commented on pkg/kubectl/cmd/set/set_volume.go:
```diff
@@ -0,0 +1,960 @@
+	}
+	if len(o.Output) > 0 || o.Local || o.DryRun {
+		for _, info := range infos {
+			if err := o.PrintObject(o.Cmd, o.Local, o.Mapper, info.Object, o.Out); err != nil {
```
if multi resources get updated, this will print just one

zjj2wry commented on pkg/kubectl/cmd/set/set_volume.go:
```diff
@@ -0,0 +1,960 @@
+		info := patch.Info
+		if patch.Err != nil {
+			failed = true
+			fmt.Fprintf(o.Err, "error: %s/%s %v\n", info.Mapping.Resource, info.Name, patch.Err)
```
should be use `utilerrors.NewAggregate(allErrs)` will better

WanLinghao commented:
@cblecker PTAL

WanLinghao commented:
#54554 rebase after this patch

WanLinghao commented on pkg/kubectl/cmd/set/set_volume.go:
```diff
@@ -0,0 +1,980 @@
+	// if a claim should be created, generate the info we'll add to the flow
+	if o.Add && o.AddOpts.CreateClaim {
+		claim := o.AddOpts.createClaim()
+		m, err := o.Mapper.RESTMapping(api.Kind("PersistentVolumeClaim"))
```
@zjj2wry @liggitt I got trouble here, can't find Kind function in "k8s.io/api/core/v1"

zjj2wry replied:
GroupVersionKind.GroupKind()

WanLinghao replied:
@zjj2wry can you describe it in detail?thanks a lot

zjj2wry replied:
`mapping, err := mapper.RESTMapping(schema.GroupKind{Kind: "PersistentVolumeClaim"})` maybe can work

WanLinghao commented:
@zjj2wry @kargakis @cblecker  reabased, PTAL

zjj2wry commented on pkg/kubectl/cmd/set/set_volume.go:
```diff
@@ -0,0 +1,980 @@
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apiserver/pkg/storage/names"
+	kcoreclient "k8s.io/client-go/kubernetes/typed/core/v1"
```
nits，use coreclient

zjj2wry commented on pkg/kubectl/cmd/set/set_volume.go:
```diff
@@ -0,0 +1,980 @@
+	"k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	kresource "k8s.io/apimachinery/pkg/api/resource"
```
what about apiresource

zjj2wry commented on pkg/kubectl/cmd/set/set_volume.go:
```diff
@@ -0,0 +1,980 @@
+		For descriptions on other volume types, see https://kubernetes.io/`)
+
+	volumeExample = templates.Examples(`
+	  # List volumes defined on all deployment in the current project
```
project is openshift define, should update it to namespace

kargakis commented on pkg/kubectl/cmd/set/set_volume.go:
```diff
@@ -0,0 +1,980 @@
+	addOpts := &AddVolumeOptions{}
+	opts := &VolumeOptions{AddOpts: addOpts}
+	cmd := &cobra.Command{
+		Use:     "volumes RESOURCE/NAME --add|--remove|--list",
```
The `--remove` functionality will likely end up in its own [`kubectl unset`](https://github.com/kubernetes/kubernetes/issues/39775) command and `--list` in [`kubectl view`](https://github.com/kubernetes/kubernetes/issues/29679). That being said,I would suggest making `--add` the default here and remove all of these flags. 

cc: @kubernetes/sig-cli-pr-reviews

zjj2wry commented on pkg/kubectl/cmd/set/set_volume.go:
```diff
@@ -0,0 +1,980 @@
+	if err != nil {
+		return err
+	}
+	mapper, _ := f.Object()
```
use o.Mapper direct

zjj2wry commented on pkg/kubectl/cmd/set/set_volume.go:
```diff
@@ -0,0 +1,980 @@
+	if len(o.Output) > 0 || o.Local || o.DryRun {
+		allErrs := []error{}
+		for _, info := range infos {
+			if err := o.PrintObject(o.Cmd, o.Local, o.Mapper, info.Object, o.Out); err != nil {
```
this maybe will occur error

zjj2wry replied:
https://github.com/kubernetes/kubernetes/issues/55407

zjj2wry replied:
and @smarterclayton doing refactor..

https://github.com/kubernetes/kubernetes/pull/55660#issuecomment-344172145

zjj2wry commented:
@WanLinghao this PR lgtm when your fix small nits.

WanLinghao commented:
@zjj2wry  @kargakis  ready

WanLinghao commented:
@liggitt PTAL

WanLinghao commented:
@zjj2wry   do you know who can help push this, thanks

WanLinghao commented:
@liggitt  PTAL  thanks

zjj2wry commented:
/retest
@kubernetes/sig-cli-pr-reviews

mengqiy commented:
Haven't look at the code closely, but I'm happy this PR doesn't introduce bad dependency.

WanLinghao commented:
@liggitt  are you ok with this patch? if so , when it  can be merged.

dixudx commented:
@WanLinghao Code freeze now for v1.9 release. And this is not a must-fix issue for v1.9.

Since it has already been labeled for v1.10, please be patient to wait for the end of release v1.9.

Thanks for your contribution.

WanLinghao commented:
@dixudx OK,thanks

cblecker commented:
/unassign

WanLinghao commented:
@smarterclayton @liggitt PTAL

thockin commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/54284)
<!-- Reviewable:end -->
